### ISSUE TITLE
Using the command line example, how can I get credentials to refresh?

### ISSUE THREAD

**Issue by lastobelus** [google/google-auth-library-ruby#64]
I'm authenticating against a REST Execution API script target.

Currently after authorizing, I can run `Google::Apis::ScriptV1::ExecutionRequest`s until the `expiration_time_millis`, and then they stop working with `"errorMessage"=>"Authorization is required to perform that action."` until I **manually** delete the tokens.json file and re-authenticate.

Is there something beyond the command line example in the readme that I need to do? I can't figure it out from browsing the code.

My end goal is to authenticate my rails app against my script target once on deployment, and have it run indefinitely.

**Comment by lastobelus** [comment]
I think I figured it out. The solution was to add this:

```
      credentials.fetch_access_token!({}) if credentials.expired?
```

at the end.

**Comment by sqrrrl** [comment]
How are you using the credentials? The client should be automatically refresh credentials for you.

**Comment by lastobelus** [comment]
@sqrrrl it will for accesses in a single process. But a new process with a new client needs to call `credentials.fetch_access_token!({}) if credentials.expired?`.

**Comment by sqrrrl** [comment]
I haven't been able to reproduce that. I have a few CLI samples that I've authorized previously and they all correctly refresh the credentials before calling the API.

If you have sample code that can reproduce please share.

**Comment by lastobelus** [comment]
this is my current code: 

```
    def authorize
      client_id = Google::Auth::ClientId.new(ENV['GAPPS_TARGET_CLIENT_ID'], ENV['GAPPS_TARGET_SECRET'])
      token_store = Google::Auth::Stores::RedisTokenStore.new(redis: $redis)


      authorizer = Google::Auth::UserAuthorizer.new(
        client_id, SCOPES, token_store)
      user_id = 'default'
      credentials = authorizer.get_credentials(user_id)
      if credentials.nil?
        url = authorizer.get_authorization_url(
          base_url: OOB_URI)
        puts "Open the following URL in the browser and enter the " +
             "resulting code after authorization"
        puts url
        code = gets
        credentials = authorizer.get_and_store_credentials_from_code(
          user_id: user_id, code: code, base_url: OOB_URI)
      end
      credentials.fetch_access_token!({}) if credentials.expired?
      credentials
    end

    def service
      @service ||= begin
        service = Google::Apis::ScriptV1::ScriptService.new
        service.client_options.application_name = APPLICATION_NAME
        service.authorization = authorize
        service
      end
    end
```

Before adding `credentials.fetch_access_token!({}) if credentials.expired?` it would stop working after a  couple hours with authorization errors. NOTE: I was also using `FileTokenStore` instead of `RedisTokenStore` before.

Having another look at the library code it seems to me that perhaps what is necessary/better is to call `credentials.apply!({})`?

**Comment by sqrrrl** [comment]
Still not able to reproduce, though I'm using another API to test instead of the script execution API. It's possible there's an issue elsewhere with the service itself.

As for calling apply!(), you shouldn't have to do that. It's called by the client with each request to inject the authorization header and it should be refreshing the token for you. And with the latest version, it should also being refreshing proactively before it expires/calls the API.

Need more debugging info in order to help.

Could you log the credentials before each call? You can sanitize the data of course, but what I'm interested in is seeing when the token refreshes, the expiration times, when it fails, etc.

**Comment by camertron** [comment]
@sqrrrl I ran into a manifestation of this issue as well. Please see #75 for details.

**Comment by sqrrrl** [comment]
See comment on the issue mentioned before. Bug is in `expires_within?` where it isn't correctly handling nil expiration times. Preserving the date when serializing/deserializing or just forcing it to some invalid value (0) will work around the issue.

Will aim to get a fix out early next week.

**Comment by sqrrrl** [comment]
Fixed in signet (0.7.3).
